src/host/_stream.cpp

```diff
@@ -1065,36 +1053,58 @@ using Microsoft::Console::VirtualTerminal::StateMachine;
 // - S_OK if successful.
 // - S_OK if we need to wait (check if ppWaiter is not nullptr).
 // - Or a suitable HRESULT code for math/string/memory failures.
-[[nodiscard]] HRESULT WriteConsoleWImplHelper(IConsoleOutputObject& context,
-                                              const std::wstring_view buffer,
-                                              size_t& read,
-                                              bool requiresVtQuirk,
-                                              std::unique_ptr<WriteData>& waiter) noexcept
+static [[nodiscard]] HRESULT WriteConsoleWImplHelper(
+    IConsoleOutputObject& context,
+    const std::wstring_view& buffer,
+    size_t& read,
+    bool requiresVtQuirk,
+    std::unique_ptr<WriteData>& waiter) noexcept
 {
-    try
+    // Set out variables in case we exit early.
+    read = 0;
+    waiter.reset();
+
+    RETURN_HR_IF(E_UNEXPECTED, buffer.size() > SIZE_T_MAX / sizeof(wchar_t));
+
+    // Convert characters to bytes to give to DoWriteConsole.
+    auto data = const_cast<wchar_t*>(buffer.data());
+    auto remaining = buffer.size();
+    NTSTATUS Status = S_OK;
+
+    do
     {
-        // Set out variables in case we exit early.
-        read = 0;
-        waiter.reset();
+        auto chunkLength = std::min<size_t>(remaining * sizeof(wchar_t), 4 * 1024 * 1024);
 
-        // Convert characters to bytes to give to DoWriteConsole.
-        size_t cbTextBufferLength;
-        RETURN_IF_FAILED(SizeTMult(buffer.size(), sizeof(wchar_t), &cbTextBufferLength));
+        LockConsole();
 
-        NTSTATUS Status = DoWriteConsole(const_cast<wchar_t*>(buffer.data()), &cbTextBufferLength, context, requiresVtQuirk, waiter);
+        try
+        {
+            Status = DoWriteConsole(data, &chunkLength, context, requiresVtQuirk, waiter);
+        }
+        catch (...)
+        {
+            Status = wil::ResultFromCaughtException();
+        }
+
+        UnlockConsole();
 
-        // Convert back from bytes to characters for the resulting string length written.
-        read = cbTextBufferLength / sizeof(wchar_t);
+        const auto chunkRead = chunkLength / sizeof(wchar_t);
+        data += chunkRead;
+        remaining -= chunkRead;
 
-        if (Status == CONSOLE_STATUS_WAIT)
+        if (FAILED_NTSTATUS(Status))
         {
-            FAIL_FAST_IF_NULL(waiter.get());
-            Status = STATUS_SUCCESS;
+            break;
         }
+    } while (remaining != 0);
 
-        RETURN_NTSTATUS(Status);
+    read = buffer.size() - remaining;
+
+    if (Status == CONSOLE_STATUS_WAIT)

```

Comment:
Okay, here's something I am concerned about. Right now, the console _cannot_ transition into a blocked state during output: if the user selects text during output, *all* or *none* of the output is written and if a waiter is enqueued it is for *all* of the output.

It looks like we are changing to allow other things to lock the console during output. That's not bad, but the waiter machinery scares me. I know the caller will not observe a difference, but... will we? What if you select text in the middle of a 4GB write? Have we comprehensively tested to make sure that we don't restart the write, and we don't buffer too much or too little, and that it does complete properly and in the right order when selection is released?

What about the case where there are two applications calling WriteConsole simultaneously? There's only one I/O dispatcher, so I have to assume that we will treat them as serial... but what if one of them gets pushed into the wait queue?

Reply:
Good point. So maybe we should only go with the `til::ticket_lock` change then?
Assuming the client is well behaved and writes in chunks of reasonable length, it would still greatly improve the responsiveness of conhost/conpty.

Reply:
Yeah, let's split these changes us

Reply:
up

Reply:
Done. 👍